Update purchase list view

fditrapani opened the pull request:
#### Changes proposed in this Pull Request

This PR updates the design of the purchases listing screen with a more presentation that offers more helpful information:

**Before**
![image](https://user-images.githubusercontent.com/6981253/99919573-394a3280-2cec-11eb-8314-142fadbb3d6d.png)

**After**
![image](https://user-images.githubusercontent.com/6981253/99920147-97c4e000-2cef-11eb-8dae-5d0811d200a0.png)

![image](https://user-images.githubusercontent.com/6981253/99919627-9645e880-2cec-11eb-8540-2178152420b6.png)

![image](https://user-images.githubusercontent.com/6981253/99920161-a8755600-2cef-11eb-8eaa-18316ecd83c3.png)


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Visit the site and account level purchases screen and test with the following products:

- Jetpack
- Monthly plan
- G Suite
- Domain
- Domain mapping
- Donations, or subscriptions from a site like Long reads or http://burlingtonbuddhist.org (note that we don't get payment method data from 3rd party sites. We'll need to add this in a separate PR)

Update store admin on one, or more products, to test the following scenarios:

- Expiring in 30 days (Should see a yellow warning)
- Expiring in 7 days (Should see a red warning)

Test with a credit card that is about to expire (this can be done while sandboxed by adding a credit card with an expiration date before the subscription end date. 

Lastly, take a look in some different browsers to see if you can spot anything.

Commit:
Initial code cleanup

Commit:
First pass at styles

Commit:
Update expiry notices

Commit:
Cleanup expiry messaging

Commit:
Add payment method icon

Commit:
Add PayPal payment method

Commit:
Fix invalid markup

Commit:
Update placeholder

a8ci18n commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -55,65 +49,95 @@ class PurchaseItem extends Component {
-				</Notice>
-			);
+		if ( purchase && isPartnerPurchase( purchase ) ) {
+			return translate( 'Managed by %(partnerName)s', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **21** times:
`translate( 'This plan is managed by %(partnerName)s.' )` *ES Score: 6*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/purchase-listing-view)

a8ci18n commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -55,65 +49,95 @@ class PurchaseItem extends Component {
+			if ( creditCardHasAlreadyExpired( purchase ) ) {
+				return (
+					<span className="purchase-item__is-error">
+						{ translate( 'Credit card expired' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **26** times:
`translate( 'Expired Credit Card' )` *ES Score: 10*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/purchase-listing-view)

a8ci18n commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -55,65 +49,95 @@ class PurchaseItem extends Component {
-					<Notice isCompact status={ status } icon="notice">
-						{ translate( 'Expires %(timeUntilExpiry)s', {
+					<span className={ expiryClass }>
+						{ translate( 'Expires %(timeUntilExpiry)s on {{span}}%(date)s{{/span}}', {
```
:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

Commit:
Manage site visibility

Commit:
Add spacing between view all

Commit:
Centralize styles and componentize header

Commit:
Rename and relocate purchases navigation

Commit:
Syncronize payment methods

Commit:
Delete old purchases header directory

Commit:
Update membership listing

Commit:
Fix purchases header on mobile

a8ci18n commented on client/me/purchases/membership-item/index.tsx:
```diff
@@ -9,86 +9,98 @@ import formatCurrency from '@automattic/format-currency';
-			</div>
-		);
-	}
+	return translate( 'Renews at %(amount)s on %(date)s', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Renews at %(amount)s on %(date)s.' )` *ES Score: 7*

fditrapani replied:
Same term but no punctuation at the end.

Commit:
Fix teamcity error

Commit:
Fix team city failure

Commit:
Get membership icons

Commit:
Cleanup translation string

Commit:
Stop memory leak

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5257359

Thank you @fditrapani for including a screenshot in the description! This is really helpful for our translators.

fditrapani commented on client/me/purchases/purchases-site/index.jsx:
```diff
@@ -28,51 +27,26 @@ import { managePurchase } from '../paths';
 
 	return (
 		<div className={ classNames( 'purchases-site', { 'is-placeholder': isPlaceholder } ) }>
 			<QuerySites siteId={ siteId } />
```
I'm not sure why `<QuerySites>` and the following `<AsyncLoad>` components are here since the data is being passed in from the parent component. If I remove these, I noticed it stops a weird jump in the order of the products from happening. Does anyone know why we have these here?

michaeldcain replied:
Is the data being passed from the parent? It looks like this make sure that the current `site` object is available in state. It's then queried in this component's `connect` function and padded directly to `PurchaseItem`.

michaeldcain replied:
The `AsyncLoad` component is rendering the `product-plan-overlap-notices` block. I don't think it's causing the order of the products to change unless it's doing its own querying.

fditrapani replied:
> Is the data being passed from the parent?

It might be other data coming in. The parent has the purchases by site. 

> product-plan-overlap-notices

What exactly are those?

michaeldcain replied:
> What exactly are those?

No idea :) looks like something related to Jetpack plans/products.

pottedmeat replied:
Can you describe what happens during the jump?

fditrapani replied:
I have a combination of products on WordPress.com and Jetpack. After the initial load, the order of products changes which gives off the appearance that something is jumping.

pottedmeat replied:
It looks like `state.purchases.data` is set by the `PURCHASES_USER_FETCH_COMPLETED` action then again by `PURCHASES_SITE_FETCH_COMPLETED`.

`client/state/purchases/reducer.js:updatePurchases` might remove some of the purchases from the first call then add them again to the end.

Commit:
Fix alignment

Commit:
Add credit card payment method for memberships

michaeldcain committed:
Cleanup

michaeldcain committed:
Add showSite default; removed unused domain prop

michaeldcain committed:
Make MembershipTerms and MembershipType functional components

- use siteId instead of url for icon query
- add parameter types

pottedmeat commented on client/me/purchases/purchases-list/index.jsx:
```diff
@@ -96,7 +97,7 @@ class PurchasesList extends Component {
 		const { subscriptions } = this.props;
 
-		if ( ! subscriptions.length ) {
+		if ( ! subscriptions.length || this.isDataLoading() ) {
```
Do you remember why this was added?

fditrapani replied:
yes, the memberships subscriptions were loading before the wordpress.com purchases. This extra condition made sure they showed up at the same time.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5257359

Thank you @fditrapani for including a screenshot in the description! This is really helpful for our translators.

Commit:
Remove faulty comment

Commit:
Fix broken test

michaeldcain reviewed (approved):
I tested this with accounts that have a number of different purchases on different site types (simple, Atomic, and Jetpack). I think this is good to go as-is and we can address any feedback in additional PRs.

Really nice work.

DavidRothstein commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -137,138 +159,145 @@ class PurchaseItem extends Component {
+			return translate( '%(purchaseType)s for {{button}}%(site)s{{/button}}', {
+				args: {
+					purchaseType: purchaseType( purchase ),
+					site: site.name,
```
This is where I'm getting the fatal error/whitescreen (there could be others too, of course, but this is the first one to trigger it): `Uncaught TypeError: Cannot read property 'name' of null at PurchaseItem.getPurchaseType`

Steps to reproduce:
1. Connect a Jetpack site.
2. Purchase a subscription on that site.
3. Disconnect the Jetpack site.

So the code probably needs to handle the empty site case better.

Interestingly, I think I sometimes also saw the same error (and same whitescreen) on certain page loads of a different account (one that didn't have a disconnected Jetpack site) but I wasn't able to reliably reproduce it there.  I wonder if there's also something going on where the `site` object isn't always loaded by the time it's passed in here.

fditrapani replied:
Good find! thanks for calling this out. I think we can check for site to display this particular message and if there isn't one, we can just show the product name.

fditrapani replied:
I've updated the component so that the disconnected messaging shows in the status column:

**Before**
![image](https://user-images.githubusercontent.com/6981253/100376628-dc959300-2fdd-11eb-81f2-4f03faf24328.png)


**After**
![image](https://user-images.githubusercontent.com/6981253/100376522-aa843100-2fdd-11eb-8172-25d0030dd779.png)

The changes are in this commit: https://github.com/Automattic/wp-calypso/pull/47610/commits/dc2885fded8f4110d3cd4fb0818c90aacf961a8f

fditrapani replied:
I also just added a custom icon for disconnected sites:

![image](https://user-images.githubusercontent.com/6981253/100387480-2ee2ae00-2ff6-11eb-8e7f-c5cab4455b2d.png)

DavidRothstein commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -137,138 +159,145 @@ class PurchaseItem extends Component {
+			const imgSrc =
+				purchase.payment.type === 'credit_card'
+					? getPaymentMethodImageURL( purchase.payment.creditCard.type )
+					: payPalImage;
```
I'm not sure we should assume that any renewable payment method which isn't a credit card must necessarily be PayPal.  (It might be currently true -- to be honest I have no idea -- but even if it is, it's an assumption that could easily break in the future.)

I think for now maybe you could only display the PayPal info if `purchase.payment.type === 'paypal'` and just let anything else display as an empty payment method for the time being?

fditrapani replied:
That's a good call @DavidRothstein. Thanks for the suggestion.

fditrapani replied:
All cleaned up in this commit: https://github.com/Automattic/wp-calypso/pull/47610/commits/f6bce851add556769574688078b0489cbafaaafb

DavidRothstein commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -137,138 +159,145 @@ class PurchaseItem extends Component {
+			const altText =
+				purchase.payment.type === 'credit_card'
+					? purchase.payment.creditCard.type
+					: purchase.payment.typ;
```
Looks like a typo here.

fditrapani replied:
Thanks for catching that!

fditrapani replied:
Fixed in this commit: https://github.com/Automattic/wp-calypso/pull/47610/commits/c5e06c98ad12478b46d5e9ad8e660d40a428e02c

DavidRothstein commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -55,65 +49,93 @@ class PurchaseItem extends Component {
 		if ( isExpiring( purchase ) ) {
-			if ( expiry < moment().add( 30, 'days' ) ) {
-				const status = isRecentMonthlyPurchase( purchase ) ? 'is-info' : 'is-error';
+			if ( expiry < moment().add( 30, 'days' ) && ! isRecentMonthlyPurchase( purchase ) ) {
```
I'm not sure this is a good change -- for example, if I have a monthly subscription with auto-renew off that expires tomorrow, shouldn't I always see a message about that?

I think it also makes things inconsistent with what you see when you click through to the subscription (I'm pretty sure we do show a message highlighting the upcoming expiration there).

fditrapani replied:
The `isRecentMonthlyPurchase` checks whether the purchase was made in the last week so if you bought a monthly plan, turned auto-renew off, and your purchase expires tomorrow, then this statement will be true and you'll see messaging consistent with the following screen..

DavidRothstein replied:
Good point -- somehow I missed that `isRecentMonthlyPurchase` was only true for the first week.  This seems OK.

DavidRothstein reviewed (changes requested):
This is looking very good overall, but it's causing a whitescreen on one of my test accounts (which has a purchase on a disconnected Jetpack site).  I'll leave more details about that in an inline code comment.

Some other things I noticed:

1. We're now showing "Credit card expiring soon" any time the credit card expires before the subscription (even if it's many months or even years from now, i.e. not soon).

    Before (no message about the credit card):
    ![old](https://user-images.githubusercontent.com/235183/100312393-36677000-2f80-11eb-9b81-5b55c876e9b1.png)

    After:
    ![new](https://user-images.githubusercontent.com/235183/100312401-3a938d80-2f80-11eb-8bd9-94e85b792bb0.png)

    I'm not sure this is a good change, especially since it replaces other important information (i.e. the message about the next renewal).  I think if you want to show the message at all in that scenario, at least it would be good to change the wording and make it look more like an informational message than a warning (and also make complement the message about the renewal date rather than replacing it).

2. Personally I miss the "Monthly subscription" and "Annual subscription" labels that have been removed from the WordPress.com plans on this page (you can see the removal in the above screenshots); I think they might give good context about what the upcoming renewal actually represents?

3. Maybe we need some handling for subscriptions without a defined purchase type.  Here's what it looks like for Jetpack Search, for example (I guess it doesn't look terrible actually, but I'm mentioning it because it doesn't look like the code expected this situation -- i.e. it seems like it always expects "Site Plan" or some other descriptive phrase to be there):
![jetpack-search](https://user-images.githubusercontent.com/235183/100312409-3ebfab00-2f80-11eb-8c1c-6ce190ac1f16.png)

Commit:
Add better support for disconnected sites

Commit:
Fix alt text

Commit:
Update payment method graphic logic

Commit:
Update credit card messaging

Commit:
Change credit card error to warning

fditrapani commented:
Thanks for the review and the thoughtful feedback @DavidRothstein!

> 1. We're now showing "Credit card expiring soon" any time the credit card expires before the subscription (even if it's many months or even years from now, i.e. not soon).

You made some good points about the credit card messaging. I've updated the copy with the renewal date and changed it to a warning.

![image](https://user-images.githubusercontent.com/6981253/100378274-c89f6080-2fe0-11eb-9b70-738783d90a0f.png)

> 2. Personally I miss the "Monthly subscription" and "Annual subscription" labels that have been removed from the WordPress.com plans on this page (you can see the removal in the above screenshots); I think they might give good context about what the upcoming renewal actually represents?

As far as I could tell, we only have that information for plans so I didn't want to inconsistently display it for some products and not others. I think it adds a nice touch and wouldn't be apposed to adding it in a future PR to the status so it reads `Renews annually at $48 on September 23, 2021`.

> 3. Maybe we need some handling for subscriptions without a defined purchase type. Here's what it looks like for Jetpack Search, for example (I guess it doesn't look terrible actually, but I'm mentioning it because it doesn't look like the code expected this situation -- i.e. it seems like it always expects "Site Plan" or some other descriptive phrase to be there):

Yeah, this accidentally worked out well. I started writing some code to handle this but found I was either duplicating code or replicating the method I'm using to retrieve this string which returns `null` if there isn't any meta data. Do you know if there are any issues with passing `null` to a translate argument?

Commit:
Add some new balance flavour

Commit:
Add icon for disconnected sites

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5257359

Thank you @fditrapani for including a screenshot in the description! This is really helpful for our translators.

DavidRothstein commented on client/me/purchases/purchase-item/index.jsx:
```diff
@@ -137,115 +208,151 @@ class PurchaseItem extends Component {
+							alt={ purchase.payment.creditCard.type }
+							className="purchase-item__payment-method-card"
+						/>
+						{ purchase.payment.type === 'credit_card' && purchase.payment.creditCard.number }
```
(minor) You don't need the `purchase.payment.type === 'credit_card'` because it was already checked above.

fditrapani replied:
Aah right! Thanks for catching that. I forgot to remove it from how I had the code written before.

DavidRothstein reviewed (commented):
Thanks -- this looks and works a lot better!

>As far as I could tell, we only have that information for plans so I didn't want to inconsistently display it for some products and not others. I think it adds a nice touch and wouldn't be apposed to adding it in a future PR to the status so it reads `Renews annually at $48 on September 23, 2021`.

That sounds good -- I added https://github.com/Automattic/wp-calypso/issues/47877 now to track that.

>I started writing some code to handle this but found I was either duplicating code or replicating the method I'm using to retrieve this string which returns null if there isn't any meta data. Do you know if there are any issues with passing null to a translate argument?

I don't know of any technical issues, but I think it could potentially affect the translation quality -- because it's possible that just leaving the first word out doesn't work anywhere nearly as nicely in some languages as it does in English.

To avoid that I think you could just do something like this?

```
const purchaseType = purchaseType( purchase );
if ( purchaseType ) {
  return translate( '%(purchaseType)s for {{button}}%(site)s{{/button}}', ... );
} else {
  return translate( 'For {{button}}%(site)s{{/button}}', ... ); // Or whatever other text you want
}
```

Commit:
Cleanup extra condition

Commit:
Add condition for missing purchase type

fditrapani commented:
Thanks for the follow up @DavidRothstein. I got the fixes in for that unnecessary condition and the translation string.

Commit:
Fix conceirge banner margins + padding

DavidRothstein reviewed (approved):
I did a final test and this seems good to me now - newest changes look fine too!

fditrapani commented:
Thanks @DavidRothstein!

Commit:
Fix linting issue

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5257359).

sirbrillig commented on client/me/purchases/membership-item/index.tsx:
```diff
@@ -1,69 +1,123 @@
+			.catch( ( err ) => setErrors( err ) );
+	}
+
+	useEffect( () => {
```
This will re-fetch the data repeatedly on every render until one of them succeeds. That might not be what you want. It's usually a good idea to provide a dependency array to `useEffect`.

fditrapani replied:
There was something weird happening and that's why I added the condition to check if the data had loaded here: https://github.com/Automattic/wp-calypso/pull/47610/files/468995d0e7e0f204b4ad8081aadc2e46b2029293#diff-f84151fb3113ed3ece0d28671295958b9b8f443b388cceef48763749aa6ee3f8R109. 

Can you please explain what you mean by providing a dependency array?

sirbrillig replied:
Yeah, the condition check is a good idea! Sorry I didn't mention this earlier!

`useEffect` with one (function) argument says, "after render, run this function".

`useEffect` can have a second argument, which is an array of variables which it calls "dependencies". If the dependency array is present, `useEffect` means, "after render, run this function, but only if the identity of one of these variables has changed". If the array is empty, it means to run only once.

So `useEffect(() => fetchData())` means "fetch data on every render". `useEffect(() => fetchData(), [isLoaded])` means "fetch data on every render that `isLoaded` has changed". `useEffect(() => fetchData(), [])` means "fetch data once when the component first renders".

So this `useEffect` here says "fetch data on every render, but don't fetch if we have completed fetching". This is a little tricky because since fetching the data is async, there could be any number of renders before `loadData` becomes false. It will re-fetch the data during each of those renders. 

Notice also that if there's an error, `loadData` remains true, so it will continue fetching on every render.

sirbrillig replied:
It's not a big bug, I just wanted to mention it as it should be cleaned up eventually. Worst case the browser will be fetching that data a lot, but everything should still work.

fditrapani replied:
Cool, yeah no problem. Happy to clean it up. Thanks for the background!

sirbrillig replied:
No worries. Nice job with this refactor, by the way! 👏

fditrapani replied:
Thank you! I got fix posted here: https://github.com/Automattic/wp-calypso/pull/48242. Let me know if that's better.
